lib/smart_proxy_remote_execution_ssh/runners/script_runner.rb

```diff
@@ -258,38 +226,61 @@ def indent_multiline(string)
-      ssh_options[:password] = @ssh_password if @ssh_password
-      ssh_options[:passphrase] = @key_passphrase if @key_passphrase
-      ssh_options[:keys_only] = true
-      # if the host public key is contained in the known_hosts_file,
-      # verify it, otherwise, if missing, import it and continue
-      ssh_options[:verify_host_key] = VERIFY_HOST_KEY
-      ssh_options[:auth_methods] = available_authentication_methods
-      ssh_options[:user_known_hosts_file] = prepare_known_hosts if @host_public_key
-      ssh_options[:number_of_password_prompts] = 1
-      ssh_options[:verbose] = settings[:ssh_log_level]
-      ssh_options[:logger] = Proxy::RemoteExecution::Ssh::LogFilter.new(Proxy::Dynflow::Log.instance)
-      return ssh_options
+      @session && @cleanup_working_dirs
+    end
+
+    # Creates session with three pipes - one for reading and two for
+    # writing. Similar to `Open3.popen3` method but without creating
+    # a separate thread to monitor it.
+    def session(args, in_stream: true, out_stream: true, err_stream: true)
+      @session = true
+
+      in_read, in_write = in_stream ? IO.pipe : '/dev/null'
+      out_read, out_write = out_stream ? IO.pipe : [nil, '/dev/null']
+      err_read, err_write = err_stream ? IO.pipe : [nil, '/dev/null']
+      command_pid = spawn(*args, :in => in_read, :out => out_write, :err => err_write, :close_others => false)
+      in_read.close if in_stream
+      out_write.close if out_stream
+      err_write.close if err_stream
+
+      return command_pid, in_write, out_read, err_read
+    end
+
+    def ssh_options(with_pty = false)
+      ssh_options = []
+      ssh_options << "-tt" if with_pty
+      ssh_options << "-o User=#{@ssh_user}"
+      ssh_options << "-o Port=#{@ssh_port}" if @ssh_port
+      ssh_options << "-o IdentityFile=#{@client_private_key_file}" if @client_private_key_file
+      ssh_options << "-o IdentitiesOnly=yes"
+      ssh_options << "-o StrictHostKeyChecking=no"
+      ssh_options << "-o PreferredAuthentications=#{available_authentication_methods.join(',')}"
+      ssh_options << "-o UserKnownHostsFile=#{prepare_known_hosts}" if @host_public_key
+      ssh_options << "-o NumberOfPasswordPrompts=1"
+      ssh_options << "-o LogLevel=#{settings[:ssh_log_level]}"
+      ssh_options << "-o ControlMaster=auto"
+      ssh_options << "-o ControlPath=#{local_command_file("socket")}"
+      ssh_options << "-o ControlPersist=yes"
     end
 
     def settings
       Proxy::RemoteExecution::Ssh::Plugin.settings
     end
 
+    def create_fd(sshpass)
+      io_r, io_w = IO.pipe
+      io_w << sshpass
+      io_w.close
+      io_r
+    end
+
+    def get_args(command, with_pty = false)
+      args = []
+      args += ['/usr/bin/sshpass', '-d', create_fd(@ssh_password).fileno.to_s] if @ssh_password
+      args += ['/usr/bin/sshpass', '-P', 'passphrase', '-d', create_fd(@key_passphrase).fileno.to_s] if @key_passphrase
```

Comment:
I'm afraid we might be leaking FDs here. I don't really see the reading end being closed after the child process is forked. It would be good to have per-run index of all FDs/pipes we've opened so that they can be closed properly